### PR TITLE
[Streamlet Scala API] Add Scala StreamletImpl Support - Part III

### PR DESCRIPTION
**Pull request by erenavsarogullari**
This PR aims the following changes:

Adding Scala Streamlet `clone`, `join` and `join with type` functions support
Adding `ScalaToJavaConverter` following transformation logics:
- Scala `Join` Function to `SerializableBiFunction`

Adding UT Coverages for both Scala `StreamletImpl` and `ScalaToJavaConverter`

Also: Scala `Streamlet` API UT cases' total execution time is as follows:
//heron/api/tests/scala:api-scala-test PASSED in 0.8s

### CODE REVIEW

File: heron/api/tests/scala/com/twitter/heron/streamlet/scala/converter/ScalaToJavaConverterTest.scala [hunk 1]
```diff
@@ -43,6 +43,17 @@ class ScalaToJavaConverterTest extends BaseFunSuite {
...
 
+  test("ScalaToJavaConverterTest should support SerializableBiFunction") {
+    def numbersToStringFunction(number1: Int, number2: Long): String =
+      (number1.toLong + number2).toString
+    val serializableBiFunction =
+      ScalaToJavaConverter.toSerializableBiFunction[Int, Long, String](
+        numbersToStringFunction)
+    assertTrue(serializableBiFunction
+      .isInstanceOf[
+        com.twitter.heron.streamlet.SerializableBiFunction[Int, Long, String]])
```
Comment (nwangtw) [inline heron/api/tests/scala/com/twitter/heron/streamlet/scala/converter/ScalaToJavaConverterTest.scala]: to import

File: heron/api/tests/scala/com/twitter/heron/streamlet/scala/impl/StreamletImplTest.scala [hunk 2]
```diff
@@ -70,4 +74,86 @@ class StreamletImplTest extends BaseFunSuite {
...
+                                        numClones: Int): Unit = {
+    val supplierStreamletImpl =
+      supplierStreamlet.asInstanceOf[StreamletImpl[R]]
+    assertEquals(numClones, supplierStreamletImpl.getChildren.size)
+    for (index <- 0 to (numClones - 1)) {
+      assertTrue(
+        supplierStreamletImpl
+          .getChildren(index)
+          .isInstanceOf[
+            com.twitter.heron.streamlet.impl.streamlets.MapStreamlet[_, _]])
```
Comment (nwangtw) [inline heron/api/tests/scala/com/twitter/heron/streamlet/scala/impl/StreamletImplTest.scala]: to import

File: heron/api/tests/scala/com/twitter/heron/streamlet/scala/impl/StreamletImplTest.scala [hunk 3]
```diff
@@ -70,4 +74,86 @@ class StreamletImplTest extends BaseFunSuite {
...
+      val mapStreamlet = supplierStreamletImpl
+        .getChildren(index)
+        .asInstanceOf[
+          com.twitter.heron.streamlet.impl.streamlets.MapStreamlet[R, R]]
```
Comment (nwangtw) [inline heron/api/tests/scala/com/twitter/heron/streamlet/scala/impl/StreamletImplTest.scala]: to import

File: heron/api/tests/scala/com/twitter/heron/streamlet/scala/impl/StreamletImplTest.scala [hunk 4]
```diff
@@ -70,4 +74,86 @@ class StreamletImplTest extends BaseFunSuite {
...
+  private def verifyJoinedStreamlet[R](supplierStreamlet: Streamlet[R],
+                                       expectedName: String,
+                                       expectedNumPartitions: Int): Unit = {
+    val supplierStreamletImpl =
+      supplierStreamlet.asInstanceOf[StreamletImpl[R]]
+    assertEquals(1, supplierStreamletImpl.getChildren.size)
+    assertTrue(supplierStreamletImpl
+      .getChildren(0)
+      .isInstanceOf[
+        com.twitter.heron.streamlet.impl.streamlets.JoinStreamlet[_, _, _, _]])
```
Comment (nwangtw) [inline heron/api/tests/scala/com/twitter/heron/streamlet/scala/impl/StreamletImplTest.scala]: to import

File: heron/api/tests/scala/com/twitter/heron/streamlet/scala/impl/StreamletImplTest.scala [hunk 5]
```diff
@@ -70,4 +74,86 @@ class StreamletImplTest extends BaseFunSuite {
...
+    val joinStreamlet = supplierStreamletImpl
+      .getChildren(0)
+      .asInstanceOf[com.twitter.heron.streamlet.impl.streamlets.JoinStreamlet[
```
Comment (nwangtw) [inline heron/api/tests/scala/com/twitter/heron/streamlet/scala/impl/StreamletImplTest.scala]: to import

**Comment by erenavsarogullari** [comment]
@nwangtw Thanks for the review again. Last comments have been addressed via last patch so it is ready for final review / merge ;)

File: heron/api/src/scala/com/twitter/heron/streamlet/scala/Streamlet.scala [hunk 6]
```diff
@@ -74,7 +77,7 @@ trait Streamlet[R] {
     *
     * @param mapFn The Map Function that should be applied to each element
     */
-  def map[T](mapFn: R => _ <: T): Streamlet[T]
+  def map[T](mapFn: R => T): Streamlet[T]
```
Comment (skanjila) [inline heron/api/src/scala/com/twitter/heron/streamlet/scala/Streamlet.scala]: why was this change needed

File: heron/api/src/scala/com/twitter/heron/streamlet/scala/impl/StreamletImpl.scala [hunk 7]
```diff
@@ -141,7 +145,14 @@ class StreamletImpl[R](
     *
     * @param numClones The number of clones to clone
     */
-  override def clone(numClones: Int): Seq[Streamlet[R]] = ???
+  override def clone(numClones: Int): Seq[Streamlet[R]] = {
```
Comment (skanjila) [inline heron/api/src/scala/com/twitter/heron/streamlet/scala/impl/StreamletImpl.scala]: can this become an perf issue if the number of clones are high?

File: heron/api/src/scala/com/twitter/heron/streamlet/scala/Streamlet.scala (hunk 6 above)
Comment (erenavsarogullari) [reply heron/api/src/scala/com/twitter/heron/streamlet/scala/Streamlet.scala]: Scala Functions support `contravariant`(parameter type) and `covariance`(return type) type parameters as default as follows :

From `scala.Function1` source code:
```
trait Function1[-T1, +R] extends scala.AnyRef {
 this : scala.Function1[T1, R] =>
  def apply(v1 : T1) : R
  ...
}
```

File: heron/api/src/scala/com/twitter/heron/streamlet/scala/impl/StreamletImpl.scala (hunk 7 above)
Comment (erenavsarogullari) [reply heron/api/src/scala/com/twitter/heron/streamlet/scala/impl/StreamletImpl.scala]: Currently, there is no limitation about the `number of clones` and `number of partitions`. I think limitation can be useful as long-term by getting team-agreement first and putting Java tier.

File: heron/api/src/scala/com/twitter/heron/streamlet/scala/impl/StreamletImpl.scala [hunk 8]
```diff
@@ -13,25 +13,28 @@
...
 import com.twitter.heron.streamlet.impl.streamlets.SupplierStreamlet
+
 import com.twitter.heron.streamlet.scala.{Sink, Streamlet}
 import com.twitter.heron.streamlet.scala.converter.ScalaToJavaConverter._
 
 object StreamletImpl {
 
-  def toScalaStreamlet[R](
-      javaStreamlet: com.twitter.heron.streamlet.Streamlet[R]): Streamlet[R] =
+  def toScalaStreamlet[R](javaStreamlet: JavaStreamlet[R]): Streamlet[R] =
```
Comment (nwangtw) [inline heron/api/src/scala/com/twitter/heron/streamlet/scala/impl/StreamletImpl.scala]: suggest to rename to fromJavaStreamlet(). The current class is streamlet.scala.impl. StreamletImpl, so I feel fromJavaStreamlet() makes more sense.

**Comment by nwangtw** [comment]
One more comment.

File: heron/api/src/scala/com/twitter/heron/streamlet/scala/impl/StreamletImpl.scala (hunk 8 above)
Comment (erenavsarogullari) [reply heron/api/src/scala/com/twitter/heron/streamlet/scala/impl/StreamletImpl.scala]: @nwangtw Thanks again for the review. Addressed.
